helix-rest/src/main/java/org/apache/helix/rest/metadatastore/TrieRoutingData.java

```diff
@@ -0,0 +1,136 @@
+    return resultMap;
+  }
+
+  public String getMetadataStoreRealm(String path) throws IllegalArgumentException {
+    TrieNode leafNode = findTrieNode(path, true);
+    return leafNode._realmAddress;
+  }
+
+  /**
+   * If findLeafAlongPath is false, then starting from the root node, find the trie node that the
+   * given path is pointing to and return it; raise IllegalArgumentException if the path does
+   * not point to any node. If findLeafAlongPath is true, then starting from the root node, find the
+   * leaf node along the provided path; raise IllegalArgumentException if the path does not
+   * point to any node or if there is no leaf node along the path.
+   * @param path - the path where the search is conducted
+   * @param findLeafAlongPath - whether the search is for a leaf node on the path
+   * @return the node pointed by the path or a leaf node along the path
+   * @throws IllegalArgumentException - when the path points to nothing or when no leaf node is found
+   */
+  private TrieNode findTrieNode(String path, boolean findLeafAlongPath)
+      throws IllegalArgumentException {
+    if (path.equals("/") || path.equals("")) {
+      if (findLeafAlongPath && !_rootNode._isLeaf) {
+        throw new IllegalArgumentException("no leaf node found along the path");
+      }
+      return _rootNode;
+    }
+
+    if (path.substring(0, 1).equals("/")) {
+      path = path.substring(1);
+    }
+    String[] splitPath = path.split("/", 0);
+
+    TrieNode curNode = _rootNode;
+    if (findLeafAlongPath && curNode._isLeaf) {
+      return curNode;
+    }
+    Map<String, TrieNode> curChildren = curNode._children;
+    for (String pathSection : splitPath) {
+      curNode = curChildren.get(pathSection);
+      if (curNode == null) {
+        throw new IllegalArgumentException("the provided path is missing from the trie");
+      }
+      if (findLeafAlongPath && curNode._isLeaf) {
+        return curNode;
+      }
+      curChildren = curNode._children;
+    }
+    if (findLeafAlongPath) {
+      throw new IllegalArgumentException("no leaf node found along the path");
+    }
+    return curNode;
+  }
+
+  /**
+   * Given a trie node, search for all leaf nodes that descend from the given trie node, and add
+   * all complete paths and realm addresses of these leaf nodes to a provided mapping. This method
+   * recursively calls itself.
+   * @param mapping - where the results (complete paths of leaf nodes and their realm addresses)
+   *          are stored
+   * @param curNode - the current trie node where the search starts
+   * @param curPath - the complete path of the current trie node
+   */
+  private static void addAllAddressesToMapping(Map<String, String> mapping, TrieNode curNode,
+      String curPath) {
+    if (curNode._isLeaf) {
+      mapping.put(curPath, curNode._realmAddress);
+      return;
+    }
+
+    for (Map.Entry<String, TrieNode> entry : curNode._children.entrySet()) {
```

Comment:
Please use level-by-level traversal for this. We try to avoid recursion in production code.

Reply:
The recursion code is now replaced with an iterative method based on the feedback from both @dasahcc and @narendly. Because of such a change, I'm storing the complete path of a trie node to the `name` field of a `TrieNode`, the reasons are as follows:
1. Storing complete paths in trie nodes makes `getAllMappingUnderPath` marginally better than storing partial paths. Without complete paths, the function has to either add/remove partial paths from a list as it operates a trie node stack, or store paths for all nodes as it operates the stack and waste memory; storing compete paths are faster, less wasteful, and simpler than both of these approaches. 
2. Trie nodes are usually seen to be associated with the complete prefix/strings they represent. Storing complete paths aligns with the common perception. 
Happy to discuss this further, but closing this conversion for now as the code is now iterative.